Add Means of Pruning Deactivated Aquifer Connections

bska opened the pull request:
This commit adds a new operation, `pruneDeactivatedCells`, at the `EclipseState` and `AquiferConfig` levels.  This is intended as a stop-gap solution to a sequencing problem that arises from `MINPV` processing.  In general we must not form connections to analytic aquifers until the set of active grid cells is fully established.

Pruning cells that become deactivated as part of `MINPV` handling nevertheless covers the common case of connection boxes applying only to the outer shell.  Solving the problem in general will however require delaying forming the analytic aquifer connections until after the active cells are fully known.

This PR supersedes and closes #2691.

bska commented:
jenkins build this please

joakim-hove commented:
Hmmm - active cells used to be a quite simple concept to grasp. Two questions/worries:

1. This is implemented as a public method in `EclipseState` - where in the setup sequence will this be called? Would it be possible to make a private method in the `EclipseState` class?
2.  How this the link to the `field_props` class - can we risk that the number of elements in the various 3D vectors now suddenly come out of sync?

Great if the current functionality actually solves the problem at hand - without creating inconsistencies with the 3D vectors - but then the main entry point in `EclipseState` should be renamed to something with `aquifer` e.g. ``EclipseState::prune_aquifer_connections()`.

blattms commented:
I would skip the (costly?) conversion to std::unordered_set either by using std::unordered_set for the public method or sorting the array (implies a copy if it is unsorted before).

bska commented:
> I would skip the (costly?) conversion to `std::unordered_set` either by using `std::unordered_set` for the public method or sorting the array (implies a copy if it is unsorted before).

The deactivated cells do indeed come in an arbitrary order so if we're basing the lookup on a sorted vector the surrounding code would be along the lines of the one below.  This is more work than forming the `unordered_set` and, arguably, more complicated 
too.  Note that `std::hash<T>` for integer types `T` is very often the identity function `hash(x) = x`.  If we're changing the interface to take an `unordered_set` directly then that affects the `MinpvProcessor` and all callers of [`CpGrid::processEclipseFormat()`](https://github.com/OPM/opm-grid/blob/40150de2a85b82eb8bfe7eebaa4c92cbabfad107/opm/grid/CpGrid.hpp#L262-L265)

```C++
void Aquancon::pruneDeactivatedCells(const std::vector<std::size_t>& deactivated)
{
    auto removed = deactivated;
    std::sort(removed.begin(), removed.end());

    for (auto& conns : this->cells) {
        auto end = std::remove_if(conns.second.begin(), conns.second.end(),
            [&removed](const AquancCell& cell) -> bool
        {
            auto pos = std::lower_bound(removed.begin(), removed.end(),
                                        cell.global_index);

            return (pos != removed.end()) && (*pos == cell.global_index);
        });

        conns.second.erase(end, conns.second.end());
    }
}
```

--- 

> This is implemented as a public method in `EclipseState` - where in the setup sequence will this be called?

Exactly once, immediately after we've established the set of cells that are deactivated due to low pore-volumes.  See PR OPM/opm-simulators#3572.

> Would it be possible to make it a private method which is called from the `EclipseState` setup?

I don't see a way to do that unless we move `MINPV` processing into the `EclipseState`.

> How does this interact with the `field_props` class - can we risk that the number of elements in the various 3D vectors now suddenly come out of sync?

See the PR.  This is called in exactly the same spot as `FieldPropsManager::reset_actnum()`.

joakim-hove commented:
> This is called in exactly the same spot as FieldPropsManager::reset_actnum().

OK - I understand; then I want a new 
```C++
void EclipseState::reset_actnum(const std::vector<int>& new_actnum, const std::vector<std::size_t>&removed_cells) {
      this->aquifer:_config.pruneDeactivedCells( deactivated );
      this->field_props.reset_actnum( new_actnum );
}
```
which can be called from the simulator.

bska commented:
> OK - I understand; then I want a new [to coalesce `reset_actnum` with `pruneDeactivatedCells()`]

It's not obvious to me that that's possible At least in current master `FieldProps::reset_actnum` is called only on rank 0 whereas `pruneDeactivatedCells()` must be called on all ranks.

joakim-hove commented:
> It's not obvious to me that that's possible

OK - I agree; that was unfortunate ;-(

blattms commented:
I think that @joakim-hove's idea of changing the method on EclipseState to indicate that this changes the aquifer only is a good one. My other comment may be ignored as noise.

bska commented:
I have pushed an update that renames the new method to `pruneDeactivatedAquiferConnections`.

bska commented:
jenkins build this please

bska commented:
jenkins build this please

bska commented:
Self-merging per prior approval.
